README.rst
========
Overview
========

Netcat for MQTT

* Free software: BSD 2-Clause License

============
Installation
============

This needs Python 3.6. At the command line::

    pip3 install mqttcat

If your system does not provide python 3.6, a more comfortable way is to use the mqtttool Docker 
container. Please see the `Github repo`_.


=============
Documentation
=============



::


    Usage: mqttcat [OPTIONS] URL

      A MQTT Message filter inspired by netcat and other Unix tools.

      Publishes Messages from STDIN to a MQTT Topic

        -or-

      Subscribes to MQTT Topic and writes messages to STDOUT

      URL - Examples:

          mqtt://hostname/topic     tcp://hostname:1883/topic
          ws://hostname/topic

      Usage Examples:

          mqttcat --echo mqtt://localhost/%23 >/dev/null

          ... will subscribe to all topics ("%23" is urlencoded #), and echo
          them to STDERR for control

          echo "Heart ... beat" | mqttcat --echo --loop --wait=3.3
          mqtt://localhost/heartbeat-topic

          ... will publish "Heart ... beat" every 3.3 seconds to the topic
          "heartbeat-topic"

          mqttcat mqtt://source/ticktock | rq 'filter (a) => { a.payload="tock"
          }' | mqttcat mqtt://destination/tock

          ... subscribe to ticktock topic on host source, filter out the
          messages whose payload is tock,     and forward those to the tock
          topic (using the rq tool https://github.com/dflemstr/rq)

    Options:
      --loglevel TEXT             Python loglevel, one of
                                  DEBUG,INFO,WARNING,ERROR,CRITICAL
      --echo / --no-echo          Echo MQTT Messages to STDERR
      --source TEXT               File to read MQTT messages to be published (use
                                  '-' for STDIN)
      --wait FLOAT                Wait time between publishing messages read from
                                  --source in seconds (can be float)
      --loop / --no-loop          Loop message publishing (starting at beginning
                                  of file after the end is reached
      --follow / --no-follow      Wait for additional in file after reaching the
                                  end (like Unix 'tail -f')
      --destination TEXT          Append JSON-encoded MQTT messages to this file
                                  (use '-' for STDOUT)
      --snapshot / --no-snapshot  Keep only the last JSON-encoded message in the
                                  file specified with --destination
      --help                      Show this message and exit.



=======
Authors
=======

* Martin Virtel - https://twitter.com/mvtango

This module is part of the `Smart Orchestra Project`_, co-financed by the 
`Federal Ministery of Economics and Technology`_.


.. _Smart Orchestra Project: http://smartorchestra.de

.. _Github repo: https://github.com/martinvirtel/docker-mqtttool

.. _Federal Ministery of Economics and Technology: https://www.bmwi.de/


CHANGELOG.rst

Changelog
=========

0.1.0 (2018-03-20)
------------------

* First release on PyPI.

0.1.1 (2018-04-13)
------------------

* Flak8-compliant

0.1.2 (2018-04-14)
------------------

* added --source and --destination options to the command line
* source file can be followed (like 'tail -f ') with --follow
* destination file can contain only the last messaget (with --snapshot)
* destinations abstracted in emittargets.py

0.1.3 (2018-04-14) (unreleased)
-------------------------------

*  made read/write mode detection more sophisticated: --source and --destination will 
   overrule the detection of detected STDIN / STDOUT redirection (this makes it
   easier to run mqttcat inside docker)
*  input and output are now compatible with JSON streams defined in RFC7464_ - those
   can be processed with jq_, for example


.. _RFC7464: https://tools.ietf.org/html/rfc7464
.. _jq: https://stedolan.github.io/jq/ 


tests/test_mqttcat.py

from click.testing import CliRunner

from mqttcat.cli import main


def test_main():
    runner = CliRunner()
    result = runner.invoke(main, [])

    assert result.output == '()\n'
    assert result.exit_code == 0


src/mqttcat/__init__.py
__version__ = '0.1.0'

import json
import logging
import sys
import time
import urllib.parse
from functools import partial
import paho.mqtt.client
import base64
from mqttcat.emittarget import AppendToFile

logger = logging.getLogger(__name__)

logging.basicConfig(level=logging.INFO, stream=sys.stderr,
                    format="%(asctime)-15s %(name)s %(levelname)s %(filename)s:%(lineno)s %(message)s")

# logging.getLogger("requests.packages.urllib3.connectionpool").setLevel(logging.CRITICAL)


def to_data(bstr):
    return "data:;base64,{}".format(base64.b64encode(bstr).encode('ascii'))


class MqttTopic(object):

    __slots__ = ('url', 'kwargs', 'client', 'topic', 'echo', 'target')

    def __init__(self, url, echo=None, target=AppendToFile(sys.stdout), **kwargs):
        parsed = urllib.parse.urlsplit(url)
        if parsed.scheme in ('ws', 'websockets'):
            kwargs["transport"] = 'websockets'
        self.kwargs = kwargs
        import paho.mqtt.client
        client = paho.mqtt.client.Client(**kwargs)
        client.connect(parsed.hostname, parsed.port if parsed.port else 1883)
        client.on_message = partial(MqttTopic.read, self)
        client.loop_start()
        self.client = client
        self.topic = urllib.parse.unquote(parsed.path[1:])
        self.echo = echo
        self.target = target
        logger.debug("Connected to {url} {kwargs}".format(**locals()))

    def subscribe(self):
        self.client.subscribe(self.topic)

    def read(self, client, userdata, message):
        obj = dict()
        for a in ("qos", "topic", "payload", "timestamp"):
            val = getattr(message, a)
            if type(val) == bytes:
                try:
                    val = val.decode("utf-8")
                except Exception:
                    val = to_data(val)
            obj[a] = val
        repres = "\x1e" + json.dumps(obj) + "\n"
        self.target.emit(repres)
        if hasattr(self.echo, 'write'):
            self.echo.write(repres)

    def publish(self, dicts, wait=None, echo=None):
        for current in dicts:
            self.client.publish(self.topic, current["payload"])
            if hasattr(echo, 'write'):
                self.echo.write(json.dumps(current)+"\n")
            if wait is True:
                raise NotImplementedError("Missing: copy wait time from input stream")
            elif float(wait) > 0:
                time.sleep(wait)

    @staticmethod
    def feeder(stream, loop=False, follow=False):
        buf = []
        for line in iter(stream.readline, None):
            logger.debug(f"Read: {line}")
            if line == '':
                if not follow:
                    break
                else:
                    time.sleep(0.2)
                    continue
            if type(line) == str:
                if line[0] == "\x1e": # ASCII RS Char, output i.e. by the jq tool with --se option
                    line = line[1:]
                try:
                    current = json.loads(line)
                except Exception as e:
                    logger.debug("'{line}' not decoded as JSON: {e}".format(**locals()))
                    current = line[:-1]
                if type(current) == str:
                    current = dict(payload=current)
            if loop:
                buf.append(current)

            yield(current)
        if loop:
            while True:
                for current in buf:
                    yield current

    def __del__(self):
        try:
            self.client.disconnect()
        except Exception as e:
            logger.debug("Error disconnecting: {e}".format(**locals()))
        else:
            try:
                self.client.loop_stop()
            except Exception as e:
                logger.debug("Error stopping loop: {e}".format(**locals()))


protocols = {
            'MQTTv31': paho.mqtt.client.MQTTv31,
            'MQTTv311': paho.mqtt.client.MQTTv311
          }


src/mqttcat/emittarget.py

class EmitTarget(object):

    __slots__ = ['args', 'kwargs', 'state']

    def __init__(self, *args, **kwargs):
        self.args = args
        self.kwargs = kwargs
        self.state = None

    def emit(self, thing):
        raise NotImplementedError


class AppendToFile(EmitTarget):

    def emit(self, thing):
        if self.state is None:
            if hasattr(self.args[0], 'write'):
                self.state = {"file": self.args[0]}
            else:
                self.state = {"file": open(*self.args, "w+", **self.kwargs)}
        self.state["file"].write(thing)
        self.state["file"].flush()

    def __del__(self):
        if self.state is not None:
            self.state["file"].close()

    def __repr__(self):
        return f"<{self.__class__.__name__} {self.args[0]}>"


class SnapshotToFile(AppendToFile):

    def emit(self, thing):
        if self.state is None:
            if hasattr(self.args[0], 'write'):
                self.state = {"file": self.args[0]}
            else:
                self.state = {"file": open(*self.args, "w+", **self.kwargs)}
        self.state["file"].seek(0)
        self.state["file"].write(thing)
        self.state["file"].flush()
        self.state["file"].truncate()


if __name__ == '__main__':
    import sys
    e = AppendToFile(sys.stdout)
    e.emit("Test")


src/mqttcat/cli.py
"""
Module that contains the command line app.

"""
import click
import sys
import logging
from mqttcat import MqttTopic
from mqttcat.emittarget import AppendToFile, SnapshotToFile

logger = logging.getLogger(__name__)


def set_root_logger(loglevel):
    try:
        if type(loglevel) == int:
            level = loglevel
        else:
            level = getattr(logging, loglevel)
            assert(type(level) == int)
        logging.getLogger().setLevel(level)
    except Exception as e:
        raise ValueError("Failed setting log level to {}: {}".format(loglevel,
                         repr(e)))


@click.command()
@click.option('--loglevel', default="INFO",
              help="Python loglevel, one of DEBUG,INFO,WARNING,ERROR,CRITICAL")
@click.option('--echo/--no-echo',
              default=False,
              help="Echo MQTT Messages to STDERR")
@click.option('--source',
              default=None,
              help="File to read MQTT messages to be published (use '-' for STDIN)")
@click.option('--wait', type=float,
              default=0.1,
              help="Wait time between publishing messages from --source in seconds (can be float)")
@click.option('--loop/--no-loop',
              default=False,
              help="Loop message publishing (starting at beginning of file after the end is reached")
@click.option('--follow/--no-follow',
              default=False,
              help="Wait for additional in file after reaching the end (like Unix 'tail -f')")
@click.option('--destination',
              default=None,
              help="Append JSON-Sequence-encoded MQTT messages to this file (use '-' for STDOUT)")
@click.option('--snapshot/--no-snapshot',
              default=False,
              help="Keep only the last JSON-encoded message in the file specified with --destination")
@click.argument('url')
def run(url, source, follow, destination, snapshot, loglevel, echo, loop, wait):
    """
    A MQTT Message filter inspired by netcat and other Unix tools.

    Publishes Messages from STDIN to a MQTT Topic

      -or-

    Subscribes to MQTT Topic and writes messages to STDOUT

    URL - Examples:

        mqtt://hostname/topic
        tcp://hostname:1883/topic
        ws://hostname/topic

    Input Format: plain text or newline-separated JSON or JSON Sequence (RFC7464)
    are all fine.

    Output Format: JSON Sequence (RFC7464)

    Usage Examples:

        mqttcat --echo mqtt://localhost/%23 >/dev/null

        ... will subscribe to all topics ("%23" is urlencoded #), and echo them to STDERR for control


        echo "Heart ... beat" | mqttcat --echo --loop --wait=3.3 mqtt://localhost/heartbeat-topic

        ... will publish "Heart ... beat" every 3.3 seconds to the topic "heartbeat-topic"


        mqttcat mqtt://source/ticktock | rq 'filter (a) => { a.payload="tock" }' | mqttcat mqtt://destination/tock

        ... subscribe to ticktock topic on host source, filter out the messages whose payload is tock,
        and forward those to the tock topic (using the rq tool https://github.com/dflemstr/rq)

        mqttcat mqtt://source/ticktock | grep 'tock' | awk 'NR%2==0' | mqttcat mqtt://destination/tock

        ... repeat every second message with 'tock'

    """
    set_root_logger(loglevel)
    # load_config(config)
    mode = None
    if echo:
        echostream = sys.stderr
    else:
        echostream = False
    if source == '-':
        sourcestream = sys.stdin
        mode = 'read'
    else:
        if source is not None:
            sourcestream = open(source)
            mode = 'read'
    if snapshot:
        tclass = SnapshotToFile
    else:
        tclass = AppendToFile
    if destination == '-':
        deststream = tclass(sys.stdout)
        mode = 'write'
    else:
        if destination is not None:
            deststream = tclass(destination)
            mode = 'write'
    if mode is None :
        if not sys.stdin.isatty():
            sourcestream = sys.stdin
            mode = 'read'
        elif not sys.stdout.isatty():
            deststream = AppendToFile(sys.stdout)
            mode = 'write'
    if mode == 'write':
        t = MqttTopic(url, echo=echostream, target=deststream)
        logger.info("Subscribed to {} - messages will be written to {}".format(url, deststream))
        t.subscribe()
        import time
        while True:
            try:
                time.sleep(10)
            except KeyboardInterrupt:
                break
    elif mode == 'read':
        t = MqttTopic(url, echo=echostream)
        logger.info("Reading from {sourcestream.name} - publishing messages to {} [loop:{loop},wait:{wait} sec.]".format(
                    url, **locals()))
        t.publish(MqttTopic.feeder(sourcestream, loop=loop, follow=follow), wait=wait, echo=echo)
    else:
        sys.stderr.write("""
        Could not determine wether to read or write.
        Use STDIN redirection or --source parameter to publish Mqtt messages.
        Use STDOUT redirection or --destination to subscribe to topic or topics.""")


if __name__ == '__main__':
    run()
